flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1119,7 +1158,11 @@ private void jobMasterFailed(JobID jobId, Throwable cause) {
 
     private void registerDispatcherMetrics(MetricGroup jobManagerMetricGroup) {
         jobManagerMetricGroup.gauge(
-                MetricNames.NUM_RUNNING_JOBS, () -> (long) jobManagerRunnerRegistry.size());
+                MetricNames.NUM_RUNNING_JOBS,
+                // metrics can be called from anywhere and therefore, have to run without the main
+                // thread safeguard being triggered. For metrics, we can afford to be not 100%
+                // accurate
```

Comment:
I'm not sure about this note. @zentol can you confirm that there is no synchronization for metric querying?

Reply:
I concluded it because [JobManagerMetricsITCase.testJobManagerMetrics](https://github.com/apache/flink/blob/c6997c97c575d334679915c328792b8a3067cfb5/flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java#L99) fails because of it...

Reply:
👍 

https://github.com/apache/flink/pull/18566#discussion_r799532804